Asset Tracker v2: Add modem module docs (#6480)
nrfconnect/sdk-nrf: applications/asset_tracker_v2/doc/modem_module.rst

.. _asset_tracker_v2_modem_module:

Modem module
############

.. contents::
   :local:
   :depth: 2

The modem module communicates with the modem to control the LTE connection and retrieve information from the modem about the device and LTE network.

Features
********

> This section documents the various features implemented by the module.

Reviewer: ```suggestion
This section documents the features implemented by the module.
```

Reviewer 2: fixed

LTE connection handling
=======================

The module uses the :ref:`lte_lc_readme` library to control and monitor the LTE connection.
The module registers an event handler with :ref:`lte_lc_readme` to receive updates from the modem about the state of the LTE connection and network information.

> If you use :ref:`liblwm2m_carrier_readme`, please refer to :ref:`modem_module_carrier_lib` for additional information on LTE connection handling.

Reviewer: ```suggestion
If you use the :ref:`liblwm2m_carrier_readme` library, refer to :ref:`modem_module_carrier_lib` for additional information on LTE connection handling.
```

Reviewer 2: fixed

> In its default configuration, the modem module will attempt to establish an LTE connection immediately when it is initialized.

Reviewer: ```suggestion
In its default configuration, the modem module makes an attempt to establish an LTE connection immediately when it is initialized.
```

Reviewer 2: fixed

There is no timeout on the connection attempt, so the modem will search for a suitable LTE network until it finds one.

> When the device is registered to an LTE network, the LTE connection is considered established, and a ``MODEM_EVT_LTE_CONNECTED`` event is sent.

Reviewer 3: All events that are sent out by modules in the app can be linked to their respective header file definitions using the suggested syntax.

> When the device is registered to an LTE network, the LTE connection is considered established, and a ``MODEM_EVT_LTE_CONNECTED`` event is sent.

Reviewer 2: Fixed

If the network registration is at some point lost, a ``MODEM_EVT_LTE_DISCONNECTED`` event is sent.

If the device connects to a new LTE cell, an ``LTE_LC_EVT_CELL_UPDATE`` is sent. The event contains cell ID and tracking area code as payload, which can be used to determine the device's coarse position.

If the modem module receives an ``APP_EVT_LTE_DISCONNECT`` event, it will instruct the modem to detach from the network, and a subsequent ``MODEM_EVT_LTE_DISCONNECTED`` can be expected when it's done.

To configure the LTE connection parameters, refer to the :ref:`lte_lc_readme` Kconfig options.

> These options can be used to select for instance which access technology to use (LTE-M or NB-IoT), default values for PSM timers and more.

Reviewer 4: ```suggestion
These options can be used to select for instance which access technology to use (LTE-M or NB-IoT), default values for power savings mode (PSM) timers and more.
```

Reviewer 2: fixed

After a successful FOTA, ``MODEM_EVT_CARRIER_REBOOT_REQUEST`` will be sent, indicating that the application should perform a graceful shutdown and reboot the device to apply the new modem firmware.
If a FOTA download fails, a ``MODEM_EVT_CARRIER_FOTA_STOPPED`` event is sent, and the application may again establish TLS connections and continue normal operation.

For more details and configuration options, you can refer to :ref:`liblwm2m_carrier_readme`.

Modem information
=================

When the application module sends out an ``APP_EVT_DATA_GET``, the modem module will check the requested data list for relevant requests:

* ``APP_DATA_MODEM_STATIC`` - Static modem data, such as configured system mode (any combination of LTE-M, NB-IoT and GNSS), ICCID, modem firmware version, application version and board version.
* ``APP_DATA_MODEM_DYNAMIC`` - Dynamic modem data, such as Cell ID, tracking area code, RSRP, IP address and PLMN (MCCMNC).
* ``APP_DATA_BATTERY`` - Voltage of the the modem's power supply.

> The module uses :ref:`modem_info_readme` to  acquire information about the modem, LTE network and the modem's power supply.

Reviewer 4: ```suggestion
The module uses :ref:`modem_info_readme` to acquire information about the modem, LTE network and the modem's power supply.
```

Reviewer 2: fixed

The response for the three different data types are sent as separate events:

> * Static data as ``MODEM_EVT_MODEM_STATIC_DATA_READY``:
> * Dynamic data as ``MODEM_EVT_MODEM_DYNAMIC_DATA_READY``:

Reviewer 4: ```suggestion
* Static data as ``MODEM_EVT_MODEM_STATIC_DATA_READY``
* Dynamic data as ``MODEM_EVT_MODEM_DYNAMIC_DATA_READY``
```

Reviewer 2: fixed

* Battery voltage as ``MODEM_EVT_BATTERY_DATA_READY``

If sampling of data fails, a corresponding error message is sent: ``MODEM_EVT_MODEM_STATIC_DATA_NOT_READY``, ``MODEM_EVT_MODEM_DYNAMIC_DATA_NOT_READY`` or ``MODEM_EVT_BATTERY_DATA_NOT_READY``.

Neighbor cell measurements
==========================

> Neighbor measurements can be requested by the application by sending an ``APP_EVT_DATA_GET`` event where ``APP_DATA_NEIGHBOR_CELLS`` is part of the requested data list.

Reviewer 4: ```suggestion
Neighbor cell measurements can be requested by the application by sending an ``APP_EVT_DATA_GET`` event where ``APP_DATA_NEIGHBOR_CELLS`` is part of the requested data list.
```

Reviewer 2: fixed

> Upon reception of that event, the modem module uses the LTE link controller to start a neighbor search of type ``LTE_LC_NEIGHBOR_SEARCH_TYPE_DEFAULT``. You can refer to :ref:`lte_lc_readme` for more details on the available search types.

Reviewer 4: ```suggestion
Upon reception of that event, the modem module uses the LTE link controller to start a neighbor cell search of type ``LTE_LC_NEIGHBOR_SEARCH_TYPE_DEFAULT``. You can refer to :ref:`lte_lc_readme` for more details on the available search types.
```

Reviewer 2: fixed

When the search completes, the module sends a ``MODEM_EVT_NEIGHBOR_CELLS_DATA_READY`` event containing the cell information received from the modem.
If the search fails, a ``MODEM_EVT_NEIGHBOR_CELLS_DATA_NOT_READY`` event is sent.

Module internals
****************

The modem module has an internal thread with a message queue for processing.
When an event is received in the :ref:`event_manager` handler, the event is converted to a message and put into the module's queue for processing in thread context.
This gives the module the flexibility to call functions that may take some time to complete.

Module states
=============

The modem module has an internal state machine with the following states:

  * ``STATE_INIT`` - The initial state of the module in which it awaits the modem to be initialized.

>     This module will enter this state only if the :ref:`liblwm2m_carrier_readme` library is enabled, as the modem is automatically initialized otherwise.

Reviewer 4: ```suggestion
    This module will enter this state only if the :ref:`liblwm2m_carrier_readme` library is enabled, because the modem is automatically initialized otherwise.
```

Reviewer 2: fixed

>   * ``STATE_DISCONNECTED`` - The module has performed all required initialization and is ready to establish an LTE connection.

Reviewer 4: ```suggestion
  * ``STATE_DISCONNECTED`` - The module has performed all required initializations and is ready to establish an LTE connection.
```

Reviewer 2: fixed

    This is the initial state for applications that do not use the :ref:`liblwm2m_carrier_readme`.
  * ``STATE_CONNECTING`` - The modem is currently searching for a suitable LTE network and attempting to establish a connection.
  * ``STATE_CONNECTED`` - The device is connected to an LTE network.
  * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request from the util module.

State transitions take place based on input from other modules via the event manager handler and the LTE link controller handler.

Configuration options
*********************

To enable the modem module, you must set the following option:

* :kconfig:`CONFIG_MODEM_MODULE`

The modem module thread stack size can be configured using the following option:

* :kconfig:`CONFIG_MODEM_THREAD_STACK_SIZE`

By default, the modem module will only send events with sampled data that has changed since the last sampling.
To send also unchanged data, the following option can be enabled:

* :kconfig:`MODEM_SEND_ALL_SAMPLED_DATA`

The module will by default automatically request Power Saving Mode (PSM) from the LTE network.

> If granted by the netwok, PSM will reduce the modem's power consumption. Note that the device is not reachable from the cloud when it is in PSM. The device exits PSM whenever the application sends data or the configured PSM TAU interval has passed.

Reviewer 4: ```suggestion
If granted by the netwok, PSM will reduce the modem's power consumption. Note that the device is not reachable from the cloud when it is in PSM. The device exits PSM whenever the application sends data or the configured PSM TAU (Tracking Area Update) interval has passed.
```

Reviewer 2: fixed

To not request PSM from the network, disable the following option:

* :kconfig:`CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES`

For more information on LTE configuration options, see :ref:`lte_lc_readme`.

Module events
*************

The :file:`asset_tracker_v2/src/events/modem_module_event.h` header file contains a list of the events sent by the modem module.

Dependencies
************

The module uses the following |NCS| libraries:

* :ref:`event_manager`
* :ref:`lte_lc_readme`
* :ref:`modem_info_readme`

API documentation
*****************

| Header file: :file:`asset_tracker_v2/src/events/modem_module_event.h`

> | Source files: :file:`asset_tracker_v2/src/events/modem_module_event.c`

Reviewer 2: added

    This is the initial state for applications that do not use the :ref:`liblwm2m_carrier_readme` library.
  * ``STATE_CONNECTING`` - The modem is currently searching for a suitable LTE network and attempting to establish a connection.
  * ``STATE_CONNECTED`` - The device is connected to an LTE network.
  * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request from the utility module.

State transitions take place based on input from other modules through the event manager handler and the LTE link controller handler.

Configuration options
*********************

You can set the following options to configure the modem module:

.. _CONFIG_MODEM_MODULE:

CONFIG_MODEM_MODULE - Configuration for modem module
   This option enables the modem module.

.. _CONFIG_MODEM_THREAD_STACK_SIZE:

CONFIG_MODEM_THREAD_STACK_SIZE -  Configuration for thread stack size
   This option configures the modem module thread stack size.

.. _CONFIG_MODEM_SEND_ALL_SAMPLED_DATA:

CONFIG_MODEM_SEND_ALL_SAMPLED_DATA - Configuration for sending all sampled data
   By default, the modem module sends only events with sampled data that has changed since the last sampling.

>    To send unchanged data also, enable CONFIG_MODEM_SEND_ALL_SAMPLED_DATA.

Reviewer: Can we just say: "..., enable this option"?

>    To send unchanged data also, enable :ref:`CONFIG_MODEM_SEND_ALL_SAMPLED_DATA <CONFIG_MODEM_SEND_ALL_SAMPLED_DATA>`.

Reviewer: Not sure whether it is reasonable to link to the Kconfig option itself.

Reviewer 2: removed the link.

.. _CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES:

CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES - Configuration for automatic requests of PSM
   The module automatically requests PSM from the LTE network.
   If PSM is granted by the network, it results in reduction of the modem's power consumption.
   Note that the device is not reachable from the cloud when it is in PSM.
   The device exits PSM whenever the application sends data, or the configured PSM TAU (Tracking Area Update) interval has passed.

>    To not request PSM from the network, disable CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES.

Reviewer: Same here, "..., disable this option".

>    To not request PSM from the network, disable :ref:`CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES <CONFIG_MODEM_AUTO_REQUEST_POWER_SAVING_FEATURES>`.

Reviewer: Also this links to the Kconfig option itself.
